packages/jest-runtime/src/index.ts

```diff
@@ -776,7 +778,46 @@ class Runtime {
       return this._environment.global.process;
     }
 
-    return require(moduleName);
+    const module = require(moduleName);
+
+    if (
+      !this._config.freezeCoreModules ||
+      this._config.freezeCoreModulesWhitelist.includes(moduleName)
+    ) {
+      return module;
+    }
+
+    if (this._coreModulesProxyCache[moduleName]) {
+      return this._coreModulesProxyCache[moduleName];
+    }
+
+    const set = (
+      target: object,
+      property: PropertyKey,
+      value: any,
+      receiver: any,
+    ) => {
+      if (typeof value !== 'function') {
```

Comment:
Here I had to also avoid caching in a case  when `value._isMockFunction` was true, otherwise `jest.spyOn` would refuse to work (i.e. `jest.spyOn(fs, "readFileSync")`)